File: String/strsort.cpp
//
// Created by 李汇川 on 2022/5/22.
//
#include <iostream>
#include <string>
#include <vector>
using namespace std;

string minNumber(vector<int> &nums);
bool compare(const string &a, const string &b);

int main() {
  vector<int> nums;
  nums.push_back(3);
  nums.push_back(32);
  nums.push_back(321);
  cout << minNumber(nums) << endl;

  return 0;
}

string minNumber(vector<int> &nums) {
  vector<string> strs;
  string res;
  for (auto num: nums) {
    strs.push_back(to_string(num));
  }
  // 排序
  sort(strs.begin(), strs.end(), compare);
  for (auto str: strs) {
    res += str;
  }
  return res;
}
bool compare(const string &a, const string &b) {
  return a + b < b + a;
}

File: String/strsort.py
import functools


class Solution:
    def compare(self, x, y):
        a, b = x + y, y + x
        if a > b:
            return 1
        elif a < b:
            return -1
        else:
            return 0

    def minNumber(self, nums):
        strs = [str(num) for num in nums]
        strs.sort(key=functools.cmp_to_key(self.compare))
        res = ""
        for i in range(0, len(strs)):
            res += str(strs[i])
        return res


if __name__ == '__main__':
    solution = Solution()
    nums = [3, 32, 321]
    print(solution.minNumber(nums))
